src/nvim/api/buffer.c

```diff
@@ -745,10 +758,34 @@ void nvim_buf_set_name(Buffer buffer, String name, Error *err)
   }
 }
 
-/// Checks if a buffer is valid
+/// Checks if a buffer is valid and loaded. When a buffer is
+/// unloaded its file contents are freed from memory and vim
+/// cannot operate on the buffer lines until it is reloaded
+/// (usually by opening the buffer again in a new window). API
+/// methods such as |nvim_buf_line_count()| and
+/// |nvim_buf_get_lines()| will be affected.
+///
+/// Buffers may have been unloaded by the |:bunload| command or
+/// the buffer's|'bufhidden'| option.
+///
+/// @param buffer Buffer handle
+/// @return true if the buffer is valid and loaded, false otherwise.
+Boolean nvim_buf_is_loaded(Buffer buffer)
+  FUNC_API_SINCE(5)
+{
+  Error stub = ERROR_INIT;
+  buf_T *buf = find_buffer_by_handle(buffer, &stub);
+  api_clear_error(&stub);
+  return buf && buf->b_ml.ml_mfp != NULL;
+}
+
+/// Checks if a buffer is valid.
+///
+/// @note Even if a buffer is valid it may have been unloaded. Use
+///       |nvim_buf_is_loaded()| to check if a buffer is loaded.
 ///
 /// @param buffer Buffer handle
-/// @return true if the buffer is valid, false otherwise
+/// @return true if the buffer is valid, false otherwise.
 Boolean nvim_buf_is_valid(Buffer buffer)
```

Comment:
Just noticed that this returns true for `0`.

Reply:
Yeah all of the `nvim_buf_*()` functions accept a buffer id of `0`, which is interpreted as "the current buffer". This feature is used in the lua tests.

Reply:
Yes but does it make sense for this function?

Reply:
@justinmk what would 0 mean instead in this case? And why is that better than curbuf?

Reply:
@bfredl I don't understand your question about curbuf. 

I'm thinking `nvim_buf_is_valid(0)` should return false, because `nvim_buf_is_valid` is for checking validity of buffer handles.

Reply:
But 0 is a valid buffer handle, and it means curbuf. Why should it mean something different for this function?

Reply:
It's a special value that only is relevant in the context of a buffer-related function. If someone uses nvim_buf_is_valid to validate a loop or something like that, then `nvim_buf_is_valid (0) => true`  might be unhelpful.

Reply:
And exactly how is this function not even _related_ to buffers? I'm not saying 1 is obviously objectively better _a priori_, just that if we are introducing a special case, and a change of behaviour of existing input at that, I want a stronger argument than "maybe false is more useful sometimes". After all, clients can already special case 0 as they want, the _only_ functional effect of changing this now is potentially breaking code that already uses this function and didn't anticipate this change.

Reply:
If neovim was ever refactored such that the `curbuf` global variable could be NULL, perhaps during early startup when neovim hasn't had a chance to open a file yet, then `nvim_buf_is_valid(0)` would be the logical way to ask neovim "Is there a current buffer yet?".

Reply:
@phodge In that hypothetical situation: yes. Then most other `nvim_buf_` functions would also unconditionally throw an error in that situation. (if we _want_ such a situation to be exposed is of course a different question, the programming model is simpler as long as we can avoid it)